Docs: Add `@generic` support for functions

Virnetch opened the pull request:
This PR adds simple support for use of `@generic` in function definitions. It also updates the `table` function definitions, as well as `mwse.loadConfig` and `mwseLogger:assert` to make use of it. As an example, this allows the return value of `mwse.loadConfig` to match the type of the given `defaultConfig`:
![image](https://github.com/user-attachments/assets/e3deb5ec-9c2c-4ba5-a15b-135f1a56a662)


The format matches the one suggested at https://github.com/MWSE/MWSE/issues/511#issuecomment-2348800098:

```lua
-- autocomplete\definitions\global\table\get.lua
return {
	type = "function",
	description = [[Gets a value in a table. If the key doesn't exist in the table, a specified default value will be returned instead.]],
	generics = {
		{ name = "keyType" },
		{ name = "valueType" },
	},
	arguments = {
		{ name = "t", type = "{ [keyType]: valueType }" },
		{ name = "key", type = "keyType", description = "The key to use to access the table." },
		{ name = "defaultValue", type = "valueType", description = "The default value if the key didn't exist in the table." },
	},
	returns = "valueType",
}
```

There is also an optional `inherits` field:

```lua
-- definitions\global\table\deepcopy.lua
return {
	type = "function",
	description = [[Copies a table's contents. All subtables will also be copied, as will any metatable.]],
	generics = {
		{ name = "tableType", inherits = "table" }
	},
	arguments = {
		{ name = "t", type = "tableType" },
	},
	valuetype = "tableType",
}
```
<br>

[`docs/function-definitions-guide.md`](https://github.com/MWSE/MWSE/blob/master/docs/function-definitions-guide.md#function-definitions) has also been updated.

---

I also noticed a few other issues that have been fixed:
* Docs for `mwseLogger:assert` has been updated to more closely match actual implementation and the built-in `assert` function
* Fixed return type for `math.isclose` incorrectly specified as a number in docs
* Added missing documentation for `table.swap`
* Changes to `misc/package/Data Files/MWSE/core/initialize.lua`:
  * Copied the annotations generated from definitions to the actual implementations for `table`, `string` and `math`, as well as `mwse.loadConfig` and `logger:assert`. This avoids vscode displaying multiple annotations for some functions.
  * Added a `fun(fileName: string): table?` type overload to `mwse.loadConfig`, so that `nil` can be dropped from the result when called with a defaultConfig argument
  * Also added two `@diagnostic disable` lines to `initialize.lua`.


closes #511

C3pa reviewed (changes requested):
I'm glad you are working on this. These changes are more than welcome. I've left some comments.

Virnetch commented:
> I'm glad you are working on this. These changes are more than welcome. I've left some comments.

Thanks, I fixed the issues you pointed out. Also added a few more changes (listed them at the bottom of the initial comment).
Another issue I noticed was that if a function with the generic type `{ [keyType]: valueType }` is called with a "class" type table, the keyType and valueType are discarded from any unions they are used in. This leads to an issue where for example `table.get` would specify just the defaultValue's type in the result:
![image](https://github.com/user-attachments/assets/c9fecb8c-43a1-4926-8f1c-97ea49e45977)

I worked around this in 3af03ad by just adding `unknown` to the result union, but maybe there's a better way to handle it?

If you can think of other places where generics would be useful, I can add them, but other than that I think this should be good to merge.

C3pa commented on docs/function-definitions-guide.md:
```diff
@@ -34,10 +34,17 @@ Each function definition is just a regular Lua table. The following fields are a
+| Field    | Type     | Description |
+| -------- | -------- | ----------- |
+| name     | `string` | The name of the generic type. This name can then be used as a type for any of the function's arguments or return values. |
+| inherits | `string` | An optional parent type for the generic. |
```
I wasn't familiar with the fact that generics support inheritance. I had to look this up in LuaLS's changelog. I give this a thumbs-up.

C3pa reviewed (approved):
I'm very glad you added annotations to the implementation. This will suppress some type-checking inaccuracies that LuaLS was picking up from the implementation site.

I'm glad you worked on this. I haven't got any additional suggestions and this provides a good foundation.
